tests/setuptab.test.js

```diff
@@ -260,116 +264,94 @@ describe('UI spec functionality', () => {
 
-    const {
-      findByLabelText, findByTestId
-    } = renderSetupFromSpec(spec, uiSpec);
-    const controller = await findByLabelText(
-      RegExp(`${spec.args.controller.name}`)
-    );
+    const { findByLabelText } = renderSetupFromSpec(spec, uiSpec);
+    const arg1 = await findByLabelText(RegExp(`${spec.args.arg1.name}`));
     const arg2 = await findByLabelText(RegExp(`${spec.args.arg2.name}`));
     const arg3 = await findByLabelText(RegExp(`${spec.args.arg3.name}`));
     const arg4 = await findByLabelText(RegExp(`${spec.args.arg4.name}`));
-    const arg5 = await findByLabelText(RegExp(`${spec.args.arg5.name}`));
-    // The 'hide' style is applied to the whole Form.Group which
-    // includes the Label and Input. Right now, the only good way
-    // to query the Form.Group node is using a data-testid property.
-    const arg3Group = await findByTestId('group-arg3');
 
     await waitFor(() => {
       // Boolean Radios should default to "false" when a spec is loaded,
       // so controlled inputs should be hidden/disabled.
       expect(arg2).toBeDisabled();
-      expect(arg3Group).toHaveClass('arg-hide');
-
-      // This input is controlled, but has an invalid option
-      // so is not actually controlled.
-      expect(arg4).toBeVisible();
-      expect(arg4).toBeEnabled();
-      // This input is not controlled.
-      expect(arg5).toBeVisible();
-      expect(arg5).toBeEnabled();
+      expect(arg3).toBeDisabled();
+      expect(arg4).toBeDisabled();
     });
     // fireEvent.change doesn't trigger the change handler but .click does
     // even though React demands an onChange handler for controlled checkbox inputs.
     // https://github.com/testing-library/react-testing-library/issues/156
-    fireEvent.click(controller, { target: { value: 'true' } });
-    // controller.click();
+    fireEvent.click(arg1, { target: { value: 'true' } });
 
-    // Now everything should be visible/enabled.
+    // Check how the state changes as we click the checkboxes
     await waitFor(() => {
       expect(arg2).toBeEnabled();
-      expect(arg2).toBeVisible();
-      expect(arg3).toBeEnabled();
-      expect(arg3).toBeVisible();
+      expect(arg3).toBeDisabled();
       expect(arg4).toBeEnabled();
-      expect(arg4).toBeVisible();
-      expect(arg5).toBeEnabled();
-      expect(arg5).toBeVisible();
+    });
+
+    fireEvent.click(arg2, { target: { value: 'true' } });
+    await waitFor(() => {
+      expect(arg2).toBeEnabled();
+      expect(arg3).toBeEnabled();
+      expect(arg4).toBeDisabled();
     });
   });
 
-  test('expect non-boolean controller can disable/hide optional inputs', async () => {
-    // Normally the UI options are loaded from a seperate spec on disk
-    // that is merged with ARGS_SPEC. But for testing, it's convenient
-    // to just use one spec. And it works just the same.
+  test('expect dropdown options can be dynamic', async () => {
+    const mockGetVectorColumnNames = jest.fn(state => {

```

Comment:
The way this is used, I'm not even sure we need a `jest.fn` instead of just a plain `function`, right? Or if you wanted to make a slightly more "real" dropdownFunction, it could be defined exactly as you have them in `ui_config.js`, but then here we would mock the return/resolve value of `getVectorColumnNames`.

Reply:
you're right, not sure why I used a `jest.fn`! For a unit test I kind of like this over mocking `server_requests.getVectorColumnNames` because I'm just trying to test that the dropdown options can change in response to some function of state, if that makes sense